new-build support for image source

bparees opened the pull request.

rhcarvalho commented on pkg/cmd/cli/cmd/newapp.go:
```diff
@@ -446,6 +446,10 @@ func setupAppConfig(f *clientcmd.Factory, out io.Writer, c *cobra.Command, args
+
+	if len(config.SourceImage) != 0 && len(config.SourceImagePath) == 0 {
+		return cmdutil.UsageError(c, "--source-image-path must be specified when --source-image is specified.")
+	}
```
Shouldn't we check the opposite as well, `--source-image-path` is set while `--source-image` is empty?

bparees replied:
yup, will add.

rhcarvalho commented on pkg/cmd/cli/cmd/newbuild.go:
```diff
@@ -109,6 +109,8 @@ func NewCmdNewBuild(fullName string, f *clientcmd.Factory, in io.Reader, out io.
 	cmd.Flags().BoolVar(&config.DryRun, "dry-run", false, "If true, do not actually create resources.")
 	cmd.Flags().BoolVar(&config.NoOutput, "no-output", false, "If true, the build output will not be pushed anywhere.")
+	cmd.Flags().StringVar(&config.SourceImage, "source-image", "", "Specify an image to use as source for the build.")
+	cmd.Flags().StringVar(&config.SourceImagePath, "source-image-path", "", "Specify the file or directory to copy from the source image and its destination in the build directory. Format: [source]:[destination-dir]")
```
All other descriptions end with a full stop.

bparees replied:
debatable whether this should have one since it could be confused as part of the example syntax, but i'll add it.

rhcarvalho commented on pkg/generate/app/app.go:
```diff
@@ -151,6 +155,25 @@ func (r *SourceRef) BuildSource() (*buildapi.BuildSource, []buildapi.BuildTrigge
+			},
+		})
+		source.Image = imgSrc
+
```
Please kill this empty line.

mfojtik commented on pkg/generate/app/sourcelookup.go:
```diff
@@ -252,6 +264,15 @@ func (r *SourceRepository) ContextDir() string {
 	return r.contextDir
 }
 
+func (r *SourceRepository) SetSourceImage(c ComponentReference) {
```
this is wip, so I assume godoc will come later on these :)

bparees replied:
sure :)

bparees commented:
@mfojtik @csrwng @rhcarvalho ok i think this is ready for review.

I will squash the commits before merging, but for now i left it as two so @csrwng can easily just review the changes i made to what he'd already done.

there are also a couple changes in here that are not strictly related to image input, they're cleaning up new-app's image search/resolve behavior.

bparees commented:
[test]

bparees commented:
[test]

rhcarvalho commented on hack/test-cmd.sh:
```diff
@@ -57,7 +57,7 @@ tests=( $(find_tests ${1:-.*}) )
 # test-cmd specific defaults
 TMPDIR="${TMPDIR:-"/tmp"}"
-BASETMPDIR="${BASETMPDIR:-${TMPDIR}/openshift-cmd}"
+export BASETMPDIR="${BASETMPDIR:-${TMPDIR}/openshift-cmd}"
```
why are we changing this?

rhcarvalho replied:
I see.. you're creating files in `test/cmd/newapp.sh` and putting them in the temp dir... I think that's a bad idea.
How about writing a `checkResult` like below?
https://github.com/openshift/origin/pull/6615/files#diff-d47c059acde3ba8455be65d7c690ac25R1132

bparees replied:
why do you think it's a bad idea?

rhcarvalho replied:
It requires cleanup, depends on disk IO, might lead to flakiness, and in general test/cmd will give you feedback much later if something got broken.
I normally afford to run unit tests locally before submitting/updating a PR, but not really running test/cmd as part of my workflow (and sure, both groups of tests are run before a PR gets merged).

The other tests we have in test/cmd produce yaml output and grep for expected text in a single line. Specially when we have other ways, I'd strive to keep test/cmd simple.

bparees replied:
i'll change it to the unit test approach, but the claim that diskIO is a potential issue is kinda silly, all our tests depend on diskIO.  I don't see anything about this approach that's likely to be particularly flaky.

and cleanup is handled in the same way the log files are cleaned up, because it's writing to the same location.

I think this is still a simpler approach than the equivalent go code, even if it's not simpler than the other bash tests in this file.

rhcarvalho commented on test/cmd/newapp.sh:
```diff
@@ -132,6 +132,21 @@ os::cmd::expect_failure_and_text 'oc new-build mysql -o yaml' 'you must specify
 os::cmd::expect_failure_and_text 'oc new-build mysql https://github.com/openshift/ruby-hello-world --binary' 'specifying binary builds and source repositories at the same time is not allowed'
+# new-build image source tests
+os::cmd::expect_failure_and_text 'oc new-build mysql --source-image centos' 'error: --source-image-path must be specified when --source-image is specified.'
+os::cmd::expect_failure_and_text 'oc new-build mysql --source-image-path foo' 'error: --source-image must be specified when --source-image-path is specified.'
```
I think go tests run much faster and reliably than these.

Specially for testing the error messages, how about replacing these with new tests in RunBuildTests in newapp_test.go?
https://github.com/openshift/origin/pull/6615/files#diff-d47c059acde3ba8455be65d7c690ac25R1132
https://github.com/openshift/origin/pull/6615/files#diff-d47c059acde3ba8455be65d7c690ac25R1200

Those newapp_test.go tests are not the best, as they rely on GitHub, but they are easier to improve than bash.
I've just removed a lot of duplicated boilerplate from RunBuildTests.

For tests that actually require CLI interaction, then test/cmd is fine.

bparees replied:
the error message tests in particular have to be run the way they are (or some other mechanism).  Doing what you propose does not work because the error messages are not thrown from the Run() path, they are handled during argument validation.

smarterclayton replied:
We should always have test-cmd tests (to cover end user experience for error conditions) and go tests (to test functional units).  I would prefer more extensive test-cmd tests than big, coupled, generic unit tests.

rhcarvalho commented on pkg/cmd/cli/cmd/newbuild.go:
```diff
@@ -109,6 +109,8 @@ func NewCmdNewBuild(fullName string, f *clientcmd.Factory, in io.Reader, out io.
 	cmd.Flags().StringVar(&config.ContextDir, "context-dir", "", "Context directory to be used for the build.")
 	cmd.Flags().BoolVar(&config.DryRun, "dry-run", false, "If true, do not actually create resources.")
 	cmd.Flags().BoolVar(&config.NoOutput, "no-output", false, "If true, the build output will not be pushed anywhere.")
+	cmd.Flags().StringVar(&config.SourceImage, "source-image", "", "Specify an image to use as source for the build.")
```
Perhaps somewhere in the output of `oc new-build --help` we should say that `--source-image` and `--source-image-path` are meant to be used together?
This could be an usage example, or a brief mention in the description of `--source-image`.

I'm proud that users know us for the quality of the contextual help in our CLI :)

csrwng replied:
:+1: We should have an example in the example text

bparees commented:
[testonlyextended][extended:core("build with image source")]

bparees commented:
@csrwng @mfojtik ready for another review.

rhcarvalho commented:
do we need to consider how changing from single image to an array might break existing user data?
(or perhaps image source hasn't been supported in any release yet?)

bparees commented:
> do we need to consider how changing from single image to an array might break existing user data?
(or perhaps image source hasn't been supported in any release yet?)

we do not, that's why we marked the non-array api totally experimental and said it would be changing to an array w/o migration in the api doc for the non-array version.  (was discussed w/ @smarterclayton)

bparees commented:
@csrwng also note that I had to revert to using indexing for the secret variable names/mount points because the imagename and the secret name may not be valid environment variable names/keys.

open to better suggestions, but nothing simple came to mind.

csrwng commented:
@bparees I can't think of anything better. Because you're building the pod with the same build object that gets passed to the builder I can't think of a case where there would be an index mismatch.

smarterclayton commented on pkg/generate/app/cmd/newapp.go:
```diff
@@ -486,6 +490,17 @@ func (c *AppConfig) search(components app.ComponentReferences) error {
 	return errors.NewAggregate(errs)
 }
 
+func (c *AppConfig) detectPartialMatches(components app.ComponentReferences) error {
```
Does this change belong in this PR?

smarterclayton replied:
What scenario would users get this?  I would expect this to be resolve resulting in partial matching

bparees replied:
it's fixing a bug with how we reported partial matches previously.  We previously only reported partial matches for image builders, we should be reporting them for any image object.

we get here in any scenario where there is an inexact match.  Again this is just a refactoring that applies this check to all image matches instead of only image matches for source-repos.

bparees replied:
where there is *only* an inexact match (no exact match)

smarterclayton commented on pkg/generate/app/cmd/newapp.go:
```diff
@@ -919,7 +973,21 @@ func (c *AppConfig) run(acceptors app.Acceptors) (*AppResult, error) {
 
-	if err := c.resolve(components); err != nil {
+	var imageComp app.ComponentReference
+	imageComp, repositories, err = c.addImageSource(repositories)
```
I don't like this being after validate - why wouldn't this just be part of validate?

rhcarvalho replied:
(asking/commenting to try to understand better new-app's code base...)

When I read "validate" I think of a method that should have no side effects and return validation errors.
I was thinking why should calling `c.validate()` also "add Image Source"?! Then looking at the godoc for validate: "validate converts all of the arguments on the config into references to objects, or returns an error".

In my head addImageSource matches the godoc for `validate`, but the name "validate" makes no sense to me in this context, is it just me?!

bparees replied:
it's not just you.  welcome to the fun that is new-app.

mfojtik replied:
fun will be rebasing my stuff on top of this PR :-)

bparees replied:
@smarterclayton near as i can tell (i got this bit from @csrwng so he may have a better reason), it's done this way because the compRef returned from resolving an imagesource argument needs to be handled specially.  We still need to "resolve" it, but we do not want to send it through the rest of the new-app logic which would, for example, create a deploymentconfig for it.

So we need to pass that compRef to resolve(), but not to any of the other logic.  

Making that happen via the validate() method and signature would be messy.  At a minimum i think we'd have to call resolve separately from within validate().

Should this whole thing be refactored?  probably. Is this the reason to do it or the PR to do it in?  i don't think so.

smarterclayton commented on pkg/generate/app/componentresolvers.go:
```diff
@@ -299,7 +300,11 @@ func resolveExact(resolver Resolver, value string) (exact *ComponentMatch, inexa
-	return match, nil, nil
+	if match.Score == 0.0 {
+		return match, nil, nil
+	} else {
```
No else

smarterclayton replied:
So an inexact match, of which I only find one, we are going to give you a message saying "hey, we only found jenkins-1 for jenkins, tell us you want to use jenkins-1" sounds irritating to end users.

bparees replied:
correct.  Because what if all we found for "jenkins" was "jenkins-ransomware-installer"?  You really want us to blindly run the closest match?  I don't.

Or for a more realistic case, what if their app requires ruby-22 but the environment has ruby-20 installed because someone screwed up, and they just specify "ruby"?

also i'm 99% certain we discussed this and you agreed that partial matches should prompt the user, not proceed.

(and it's not new behavior being introduced in this PR, though this PR is making the behavior more consistent and better implemented)

smarterclayton commented on pkg/generate/app/pipeline.go:
```diff
@@ -237,6 +237,15 @@ func (p *Pipeline) Objects(accept, objectAccept Acceptor) (Objects, error) {
 			objects = append(objects, build)
 		}
+		if p.Build.Source != nil && p.Build.Source.SourceImage != nil && accept.Accept(p.Build.Source.SourceImage) {
+			srcImage, err := p.Build.Source.SourceImage.ImageStream()
```
Does this have to be an image stream?  I don't think so - I think this should require p.Build.Source.SourceImage.AsImageStream

bparees replied:
@csrwng did this part so he might know the intent better..   but i wouldn't think it has to be.  What does AsImageStream actually mean/check?  Confirms we were able to create an imagestream for the image reference?

bparees commented:
addressed @smarterclayton's comments (the ones that aren't still being debated, anyway)


also added a describer for imagesource..I don't love the output but i'm open to suggestions:

```
$ oc describe bc imagedockerbuild
Name:		imagedockerbuild
Created:	6 minutes ago
Labels:		build=imagedockerbuild
Latest Version:	Never built
Strategy:	Source
URL:		https://github.com/openshift/ruby-hello-world.git
Files from:	openshift/someimage:latest
 - /foo/bar/file1.zip -> dest1/dir
 - /other/dir/file2.zip -> dest2/dir
Files from:	openshift/jenkins:latest
 - /usr/lib/jenkins/jenkins.war -> injected/dir
 - /other/file.zip -> other/dir
From Image:	ImageStreamTag openshift/ruby:latest
Force Pull:	yes
Output to:	ImageStreamTag imagedockerapp:latest
Triggered by:	<none>
```

bparees commented:
[test]

bparees commented:
@smarterclayton @csrwng I think this is ready for final review/merge.

csrwng commented on pkg/cmd/cli/cmd/newbuild.go:
```diff
@@ -48,7 +48,10 @@ You can use '%[1]s status' to check the progress.`
-  $ %[1]s new-build https://github.com/openshift/ruby-hello-world RACK_ENV=development`
+  $ %[1]s new-build https://github.com/openshift/ruby-hello-world RACK_ENV=development
+
+  # Create a build config that gets its input from a remote repository and another docker image
```
really minor nit... s/docker/Docker/

csrwng commented on pkg/cmd/cli/cmd/newbuild.go:
```diff
@@ -48,7 +48,10 @@ You can use '%[1]s status' to check the progress.`
+  $ %[1]s new-build https://github.com/openshift/ruby-hello-world RACK_ENV=development
+
+  # Create a build config that gets its input from a remote repository and another docker image
+  $ %[1]s new-build https://github.com/openshift/ruby-hello-world --image-source openshift/jenkins-1-centos7 --image-source-path /var/lib/jenkins:/tmp`
```
--image-source=openshift/jenkins1-centos7 --image-source-path=/var/lib/jenkins:/tmp

csrwng commented on pkg/build/api/validation/validation.go:
```diff
@@ -225,18 +226,39 @@ func validateGitSource(git *buildapi.GitBuildSource) fielderrors.ValidationError
-	if len(imageSource.Paths) == 0 {
-		allErrs = append(allErrs, fielderrors.NewFieldRequired("paths"))
+	for _, imageSource := range imageSources {
+		allErrs = append(allErrs, validateFromImageReference(&imageSource.From).Prefix("from")...)
```
the prefix doesn't indicate which ImageSource in the list the error is coming from, use IndexPrefix(i) before .Prefix("from")

csrwng commented on pkg/build/api/validation/validation.go:
```diff
@@ -225,18 +226,39 @@ func validateGitSource(git *buildapi.GitBuildSource) fielderrors.ValidationError
+	for _, imageSource := range imageSources {
+		allErrs = append(allErrs, validateFromImageReference(&imageSource.From).Prefix("from")...)
+		if imageSource.PullSecret != nil {
+			allErrs = append(allErrs, validateSecretRef(imageSource.PullSecret).Prefix("pullSecret")...)
```
Use PrefixIndex

csrwng commented on pkg/build/api/validation/validation.go:
```diff
@@ -225,18 +226,39 @@ func validateGitSource(git *buildapi.GitBuildSource) fielderrors.ValidationError
+			allErrs = append(allErrs, validateSecretRef(imageSource.PullSecret).Prefix("pullSecret")...)
+		}
+		if len(imageSource.Paths) == 0 {
+			allErrs = append(allErrs, fielderrors.NewFieldRequired("paths"))
```
Use PrefixIndex

csrwng commented on pkg/build/api/validation/validation.go:
```diff
@@ -225,18 +226,39 @@ func validateGitSource(git *buildapi.GitBuildSource) fielderrors.ValidationError
+		if len(imageSource.Paths) == 0 {
+			allErrs = append(allErrs, fielderrors.NewFieldRequired("paths"))
+		}
+		allErrs = append(allErrs, validateImageSourcePaths(imageSource.Paths).Prefix("paths")...)
```
Alternatively, you can iterate over the list in the calling function and leave this method as is.

csrwng commented on pkg/build/api/validation/validation.go:
```diff
@@ -225,18 +226,39 @@ func validateGitSource(git *buildapi.GitBuildSource) fielderrors.ValidationError
+	for i, p := range paths {
+		if len(p.SourcePath) == 0 {
+			allErrs = append(allErrs, fielderrors.NewFieldRequired(fmt.Sprintf("[%d].sourcePath", i)))
+		}
```
We should also check that the source path is absolute

csrwng commented on pkg/cmd/cli/describe/describer.go:
```diff
@@ -214,7 +214,12 @@ func describeBuildSpec(p buildapi.BuildSpec, out *tabwriter.Writer) {
 	}
-
+	for _, image := range p.Source.Images {
+		formatString(out, "Files from", fmt.Sprintf("%s/%s", image.From.Namespace, image.From.Name))
```
may want to use something like https://github.com/openshift/origin/blob/master/pkg/util/httpproxy/upgradeawareproxy.go#L79-L90 to join namespace and name, otherwise for most cases, you'll have "/imageName" which doesn't look good. I see we could use a function like that elsewhere in this file:
https://github.com/openshift/origin/blob/master/pkg/cmd/cli/describe/describer.go#L228-L232
https://github.com/openshift/origin/blob/master/pkg/cmd/cli/describe/describer.go#L257-L261
https://github.com/openshift/origin/blob/master/pkg/cmd/cli/describe/describer.go#L278-L284
https://github.com/openshift/origin/blob/master/pkg/cmd/cli/describe/describer.go#L300-L306

bparees replied:
that function wouldn't actually help here.  it would just go through the "neither has a slash, concatenate with a slash between them" path.

will create a new helper.

csrwng commented on pkg/generate/app/cmd/newapp_test.go:
```diff
@@ -1210,6 +1211,116 @@ func TestRunBuilds(t *testing.T) {
+					"https://github.com/openshift/ruby-hello-world",
+				},
+				SourceImage:     "centos/mongodb-26-centos7",
+				SourceImagePath: "/src:/dst",
```
wouldn't this generate an invalid ImageSource ?

bparees replied:
it would, but this test doesn't actually care about that since it's not actually creating the BuildConfig.  i'll fix it to make it valid.

csrwng commented on pkg/generate/app/cmd/newapp_test.go:
```diff
@@ -1210,6 +1211,116 @@ func TestRunBuilds(t *testing.T) {
+				Components:      []string{"centos/mysql-56-centos7"},
+				To:              "outputimage",
+				SourceImage:     "centos/mongodb-26-centos7",
+				SourceImagePath: "/src:/dst",
```
same here (absolute dest)

csrwng commented on pkg/generate/app/dockerimagelookup.go:
```diff
@@ -84,7 +84,7 @@ func (r DockerClientSearcher) Search(terms ...string) (ComponentMatches, error)
 					for i := range tags {
 						tags[i].LocalOnly = true
-						glog.V(5).Infof("Found local docker image match %q for %q", tags[i].Value, term)
+						glog.V(5).Infof("Found local docker image match %q with score %q", tags[i].Value, tags[i].Score)
```
why do we need to quote the score?

bparees replied:
we don't, fixed to %f.

smarterclayton commented:
I'd prefer that when you have a single source image, it's all on one line.
That's the common case and the nested format is going to be rare.

On Mon, Jan 18, 2016 at 2:27 PM, Ben Parees <notifications@github.com>
wrote:

> addressed @smarterclayton <https://github.com/smarterclayton>'s comments
> (the ones that aren't still being debated, anyway)
>
> also added a describer for imagesource..I don't love the output but i'm
> open to suggestions:
>
> $ oc describe bc imagedockerbuild
> Name:       imagedockerbuild
> Created:    6 minutes ago
> Labels:     build=imagedockerbuild
> Latest Version: Never built
> Strategy:   Source
> URL:        https://github.com/openshift/ruby-hello-world.git
> Files from: openshift/someimage:latest
>  - /foo/bar/file1.zip -> dest1/dir
>  - /other/dir/file2.zip -> dest2/dir
> Files from: openshift/jenkins:latest
>  - /usr/lib/jenkins/jenkins.war -> injected/dir
>  - /other/file.zip -> other/dir
> From Image: ImageStreamTag openshift/ruby:latest
> Force Pull: yes
> Output to:  ImageStreamTag imagedockerapp:latest
> Triggered by:   <none>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/6546#issuecomment-172628896>.
>

csrwng commented on pkg/cmd/cli/describe/describer.go:
```diff
@@ -172,6 +172,12 @@ type BuildConfigDescriber struct {
 	host string
 }
 
+func concateNameAndNamespace(ns, name string) string {
```
s/concate/concatenate/  or simply namespacedName ?

bparees replied:
doh.  fixed.

csrwng commented:
LGTM

bparees commented:
@smarterclayton 

> I'd prefer that when you have a single source image, it's all on one line. That's the common case and the nested format is going to be rare.

that's going to be a really long line.  How do you feel about this format?
`Files from: openshift/someimage:latest [/foo/bar/file1.zip:dest1/dir,/other/dir/file2.zip:dest2/dir]`

(and just having multiple "Files from" lines if you have multiple images?

it's going to wrap depending on your path length+number of paths though.

bparees commented:
or 
`Files from: openshift/someimage:latest[/foo/bar/file1.zip->dest1/dir,/other/dir/file2.zip->dest2/dir]`

bparees commented:
@openshift/api-review sign off needed.
@liggitt this is the other PR that has significant API impact.  i really don't want to have to hold it until the middle of next week.

liggitt commented:
/me concedes defeat

mfojtik commented:
LGTM

smarterclayton commented:
If within ~80 we won't wrap, it's generally preferable to try to summarize.  I don't like the [ ] format - I don't think that adds anything, and would go for more human readable:

    Image Source: copies /var/run/foo from foo/bar:baz to /baz

bparees commented:
Hard to imagine that a docker pull spec plus 2 paths plus human readable wording is not going to wrap 80 chars on a regular basis.
```
$ wc
Image Source: openshift/jenkins-1-centos7 copies /var/lib/jenkins/foo to /tmp/foo
      1       7      82
```

bparees commented:
(I screwed up the wording you proposed, but you get the idea...it's long)

bparees commented:
Also even with a single source image you may have multiple paths you are copying from, which your format doesn't lend to displaying efficiently.  (Would we just have a line per path, repeating the image information each time?)

smarterclayton commented:
My gut is 90% of people have one image and one directory.   The format for
that can be different from the format for multiple directories and multiple
dirs.  I just don't want the common case to waste space.

On Fri, Jan 22, 2016 at 12:11 PM, Ben Parees <notifications@github.com>
wrote:

> Also even with a single source image you may have multiple paths you are
> copying from, which your format doesn't lend to displaying efficiently.
> (Would we just have a line per path, repeating the image information each
> time?)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/6546#issuecomment-173980918>.
>

bparees commented:
@smarterclayton ok, describe updated. ptal.

smarterclayton commented:
LGTM, squash and merge

bparees commented:
[merge]

bparees commented:
@smarterclayton thanks

bparees commented:
conflict fixed. [merge]

bparees commented:
[merge]

bparees commented:
flake [merge]

liggitt commented on pkg/build/api/validation/validation.go:
```diff
@@ -253,9 +255,32 @@ func validateImageSource(imageSource *buildapi.ImageSource) fielderrors.Validati
+	if filepath.IsAbs(imagePath.DestinationDir) {
+		allErrs = append(allErrs, fielderrors.NewFieldInvalid("destinationDir", imagePath.DestinationDir, "must be a relative path"))
+	}
+	if strings.HasPrefix(path.Clean(imagePath.DestinationDir), "..") {
```
is `path.Clean` system dependent? Could "foo\..\..\..\....." trick it?

rhcarvalho replied:
It's not system dependent: https://golang.org/pkg/path/.

It has no effect in backslash separated strings (e.g. Windows paths): http://play.golang.org/p/Retd8NjjXj
IOW, the cleaned version of `"foo\..\..\..\....."` is `"foo\..\..\..\....."` itself.

liggitt commented on pkg/build/builder/source.go:
```diff
@@ -256,16 +268,43 @@ func copyImageSource(dockerClient DockerClient, containerID, sourceDir, destDir
-			dockerClient.PullImage(docker.PullImageOptions{Repository: image}, dockerAuth)
+	dockerAuth := docker.AuthConfiguration{}
+	if imageSecretIndex != -1 {
+		pullSecret := os.Getenv(fmt.Sprintf("%s%d", dockercfg.PullSourceAuthType, imageSecretIndex))
```
would be handy to look at the specific envvar (`PULL_SOURCE_DOCKERCFG_PATH_<N>`), then fall back to a general envvar (`PULL_SOURCE_DOCKERCFG_PATH`) if unspecified... to avoid needing to set the same value into N envvars if you had N source images coming from the same registry
